sycl/plugins/level_zero/pi_level0.cpp

```diff
@@ -0,0 +1,3732 @@
+#include <string>
+#include <thread>
+
+#include <level_zero/zet_api.h>
+
+// Controls L0 calls serialization to w/a L0 driver being not MT ready.
+// Recognized values (can be used as a bit mask):
+enum {
+  ZeSerializeNone =
+      0, // no locking or blocking (except when SYCL RT requested blocking)
+  ZeSerializeLock = 1, // locking around each ZE_CALL
+  ZeSerializeBlock =
+      2, // blocking ZE calls, where supported (usually in enqueue commands)
+};
+static pi_uint32 ZeSerialize = 0;
+
+// This class encapsulates actions taken along with a call to L0 API.
+class ZeCall {
+private:
+  // The global mutex that is used for total serialization of L0 calls.
+  static std::mutex GlobalLock;
+
+public:
+  ZeCall() {
+    if ((ZeSerialize & ZeSerializeLock) != 0) {
+      GlobalLock.lock();
+    }
+  }
+  ~ZeCall() {
+    if ((ZeSerialize & ZeSerializeLock) != 0) {
+      GlobalLock.unlock();
+    }
+  }
+
+  static ze_result_t check(ze_result_t ZeResult, const char *CallStr,
+                           bool TraceError = true);
+
+  // The non-static version just calls static one.
+  ze_result_t checkThis(ze_result_t ZeResult, const char *CallStr,
+                        bool TraceError = true) {
+    return ZeCall::check(ZeResult, CallStr, TraceError);
+  }
+};
+std::mutex ZeCall::GlobalLock;
+
+// Controls L0 calls tracing in zePrint.
+static bool ZeDebug = false;
+
+static void zePrint(const char *Format, ...) {
+  if (ZeDebug) {
+    va_list Args;
+    va_start(Args, Format);
+    vfprintf(stderr, Format, Args);
+    va_end(Args);
+  }
+}
+
+// TODO:: In the following 4 methods we may want to distinguish read access vs.
+// write (as it is OK for multiple threads to read the map without locking it).
+
+pi_result _pi_mem::addMapping(void *MappedTo, size_t Offset, size_t Size) {
+  std::lock_guard<std::mutex> Lock(MappingsMutex);
+  auto Res = Mappings.insert({MappedTo, {Offset, Size}});
+  if (Res.second) {
+    zePrint("piEnqueueMemBufferMap: duplicate mapping detected\n");
+    return PI_INVALID_VALUE;
+  }
+  return PI_SUCCESS;
+}
+
+pi_result _pi_mem::removeMapping(void *MappedTo, Mapping &MapInfo) {
+  std::lock_guard<std::mutex> Lock(MappingsMutex);
+  auto It = Mappings.find(MappedTo);
+  if (It == Mappings.end()) {
+    zePrint("piEnqueueMemUnmap: unknown memory mapping\n");
+    return PI_INVALID_VALUE;
+  }
+  MapInfo = It->second;
+  Mappings.erase(It);
+  return PI_SUCCESS;
+}
+
+ze_result_t
+_pi_context::getFreeSlotInExistingOrNewPool(ze_event_pool_handle_t &ZePool,
+                                            size_t &Index) {
+  // Maximum number of events that can be present in an event ZePool is captured
+  // here. Setting it to 256 gave best possible performance for several
+  // benchmarks.
+  static const char *MaxNumEventsPerPoolEnv =
+      std::getenv("ZE_MAX_NUMBER_OF_EVENTS_PER_EVENT_POOL");
+  static const pi_uint32 MaxNumEventsPerPool =
+      (MaxNumEventsPerPoolEnv) ? std::atoi(MaxNumEventsPerPoolEnv) : 256;
```

Comment:
`std::atoi` will return 0 if no conversion can be performed. Are we OK with that? I see that somewhere later we do a `-1`  with this number which may lead to very large maximum number of events.

Reply:
How does `std::atoi` process a NULL pointer? Does it throw an exception or just segfaults?

Reply:
So it looks like there is a `static` just to avoid calling `std::getenv` twice.
What about
```c++
static const pi_uint32 MaxNumEventsPerPool = [] {
  const autoMaxNumEventsPerPoolEnv =
    std::getenv("ZE_MAX_NUMBER_OF_EVENTS_PER_EVENT_POOL");
  return MaxNumEventsPerPoolEnv ? std::atoi(MaxNumEventsPerPoolEnv) : 256;
}();
```

Reply:
Applied the suggestion. @s-kanaev We call atoi only if pointer is not null in this code.
Thanks for cathing the case when std::atoi will return 0, handled this case.